aten/src/ATen/native/NamedTensor.cpp

```diff
@@ -89,6 +89,38 @@ static std::vector<int64_t> aligned_size(
   return expanded_sizes;
 }
 
+Tensor as_named(const Tensor& self, DimnameList names) {
+  const auto self_names = self.names();
+  TORCH_CHECK(self_names.size() == names.size(),
+      "as_named: cannot coerce Tensor", self_names, " to Tensor", names,
+      " because they have a different number of dims (",
+      self_names.size(), " and ", names.size(), " respectively).");
+  check_names_valid_for(self, names);
+
+  for (size_t idx = 0; idx < self_names.size(); idx++) {
+    const auto& self_name = self_names[idx];
+    const auto& out_name = names[idx];
+    if (self_name == out_name || self_name.is_wildcard()) {
+      continue;
+    }
+    if (out_name.is_wildcard()) {
+      TORCH_CHECK(false,
+          "as_named: cannot coerse Tensor", self_names, " to Tensor", names,
```

Comment:
what's the use-case for a function that must not allow re-naming vs just having a rename function?

Reply:
Users can use `as_named` to assert names on a tensor. If they want to check that the names are `NCHW` they can use `tensor.as_named('N', 'C', 'H', 'W')`.

For example,

```
def conv(input):
    input = input.renamed('N', 'C', H', 'W')
    # do things
```
`renamed` is not the right function here because the input might have had names `('N', 'H', 'W', 'C')`.

Reply:
Since it's a new function we probably need docs for it (e.g. add_docstr in _torch_docs.py).

It's a bit confusing to me if as_named is meant to be an assertion more than a rename.

If you feel confident this is important to have that's fine, but in general let's bias towards adding functions / name-support as minimally as possible. Post-release it will be easier to add more than to remove them or continue to support them if we find any that aren't necessary.

Reply:
Named tensor docs will come later in a batch. I'm trying to get a demo up and running asap.

`as_named` is meant to coerce unnamed, partially named, and fully named tensors to having the desired set of names positionally. As a side effect, one can use it to assert on named tensors. I am confident we need this functionality (it's been around since the beginning of the design phase) but it's possible that we will end up splitting it into two functions or renaming it.

Reply:
Perhaps a better name would be `refine_names`?